xarray/core/indexing.py

```diff
@@ -177,18 +206,36 @@ def convert_label_indexer(index, label, index_name='', method=None,
             if np.any(indexer < 0):
                 raise KeyError('not all values found in index %r'
                                % index_name)
-    return indexer
+    return indexer, new_index
 
 
 def remap_label_indexers(data_obj, indexers, method=None, tolerance=None):
     """Given an xarray data object and label based indexers, return a mapping
-    of equivalent location based indexers.
+    of equivalent location based indexers. Also return a mapping of pandas'
+    single index objects returned from multi-index objects.
     """
     if method is not None and not isinstance(method, str):
         raise TypeError('``method`` must be a string')
-    return dict((dim, convert_label_indexer(data_obj[dim].to_index(), label,
-                                            dim, method, tolerance))
-                for dim, label in iteritems(indexers))
+
+    pos_indexers, new_indexes = {}, {}
+    for dim, label in iteritems(indexers):
+        index = data_obj[dim].to_index()
+
+        if isinstance(index, pd.MultiIndex):
+            # set default names for multi-index unnamed levels so that
+            # we can safely rename dimension / coordinate later
```

Comment:
`PandasIndexAdapter` is really for handling aspects of pandas indexes that we can't change by converting into a new index. For example, PeriodIndex claims to have `dtype=int`, when it's really a sort of object dtype.

`pandas.Index` is actually immutable, so we don't need to worry about changing the data. Calling `.copy()` just copies the metadata and reuses the same data. That said, calling `.set_names()` (which returns a new MultiIndex) is probably a more obvious way of handling this.

It does looks like we already some similar logic in `to_index()`, so that could be a reasonable place to put this, too.